### PR TITLE
KOGITO-2914: connect Trusty service and Explainability service

### PR DESCRIPTION
**Pull request by kostola**
## General description

This PR includes multiple tickets because the are all correlated to close the end to end loop ([involved component]):
- [KOGITO-3086](https://issues.redhat.com/browse/KOGITO-3086): implement a remote `PredictionProvider` that uses `/predict` API to calculate explainability. To do that, refactor `explainability-core` to support not blocking execution [explainability-addon][explainability-core]
- [KOGITO-2944](https://issues.redhat.com/browse/KOGITO-2944): implement explainability DTOs for input and output and connect them to `explainability-core` [trusty-service][explainbility-service]
- [KOGITO-2945](https://issues.redhat.com/browse/KOGITO-2945): store and read explainability result inside `trusty-service` [trusty-service]
- [KOGITO-2914](https://issues.redhat.com/browse/KOGITO-2914): expose explainability information as REST featureImportance endpoint [trusty-service]

## Description of this PR

Most of the changes are not in this repo.
This PR contains:
- Extract `predict-api` as independent addon for flexibility (btw `tracing` addon depends on `explainability` addon). We have now `explainability-quarkus-addon` and `explainability-springboot-addon` to just get `/predict` API
- Create `typedvalue-api` module that contains DTO for value serialization. They are now used by `TraceEvent`s (`tracing-addon` + `trusty-service`) and by explainability events too

## List of PRs
- https://github.com/kiegroup/kogito-runtimes/pull/720
- https://github.com/kiegroup/kogito-apps/pull/410

### CODE REVIEW

File: addons/explainability/explainability-addon-common/src/test/java/org/kie/kogito/explainability/ExplainabilityServiceTest.java [hunk 1]
```diff
@@ -54,42 +49,37 @@
 
     @Test
     public void testPerturbedExecution() {
```
Comment (r00ta) [inline addons/explainability/explainability-addon-common/src/test/java/org/kie/kogito/explainability/ExplainabilityServiceTest.java:51]: Given that with these changes multiple requests are embedded in just one http call, one request might be invalid and raise an uncaught exception. I think we have to handle this case and add a test for that, wdyt?

File: addons/explainability/pom.xml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        <groupId>org.kie.kogito</groupId>
+        <version>1.0.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>explainability</artifactId>
+    <name>Kogito :: Add-Ons :: Explainability</name>
+
+    <packaging>pom</packaging>
+
```
Comment (r00ta) [inline addons/explainability/pom.xml:16]: remove newlines

File: addons/explainability/explainability-addon-common/pom.xml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    <parent>
+        <artifactId>explainability</artifactId>
+        <groupId>org.kie.kogito</groupId>
+        <version>1.0.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>explainability-addon-common</artifactId>
+    <name>Kogito :: Add-Ons :: Explainability :: Addon :: Common</name>
+
```
Comment (r00ta) [inline addons/explainability/explainability-addon-common/pom.xml:14]: remove newline

**Review by r00ta — CHANGES_REQUESTED** [review]
Overall looks good to me, the only major comment I have is about the exception handling. Since with this new changes one of the many requests in the http body can fail we should handle it and do not return a `BAD_REQUEST`: I think that a perturbed request might contain invalid data for the model in many cases.

**Comment by r00ta** [comment]
Another side comment is that I'm not sure if `typedvalue-api` is a very informative name for a module. Wdyt?

File: addons/explainability/explainability-quarkus-addon/src/test/java/org/kie/kogito/explainability/QuarkusExplainableResourceIT.java [hunk 4]
```diff
@@ -45,20 +44,23 @@ public static ConfigBean configBeanProducer() {
 
     @Test
     @SuppressWarnings("unchecked")
-    void testRequestIsProcessedAndCorrect() throws JsonProcessingException {
+    void explainServiceTest() throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (danielezonca) [reply addons/explainability/explainability-quarkus-addon/src/test/java/org/kie/kogito/explainability/QuarkusExplainableResourceIT.java]: Done, added a test with 2 elements and on with no element

File: addons/explainability/explainability-springboot-addon/pom.xml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    <parent>
+        <artifactId>explainability</artifactId>
+        <groupId>org.kie.kogito</groupId>
+        <version>1.0.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>explainability-springboot-addon</artifactId>
+    <name>Kogito :: Add-Ons :: Explainability :: SpringBoot Addon</name>
+
```
[the remark this answers is not in the document]
Comment (danielezonca) [reply addons/explainability/explainability-springboot-addon/pom.xml]: Done

File: addons/explainability/pom.xml (hunk 2 above)
Comment (danielezonca) [reply addons/explainability/pom.xml]: Done

File: addons/explainability/explainability-addon-common/pom.xml (hunk 3 above)
Comment (danielezonca) [reply addons/explainability/explainability-addon-common/pom.xml]: Done

**Comment by kostola** [comment]
> Another side comment is that I'm not sure if `typedvalue-api` is a very informative name for a module. Wdyt?

What is your suggestion?

**Comment by r00ta** [comment]
> > Another side comment is that I'm not sure if `typedvalue-api` is a very informative name for a module. Wdyt?
> 
> What is your suggestion?

Hi @kostola , what is the purpose of such module? Is it just including the classes for the dto/event? We might go for something like `tracing-messaging-api`

**Comment by danielezonca** [comment]
@r00ta 
We are trying to avoid `tracing-` in the name because this module is also used by `explainability-service` as DTO for values.
We would like to keep services independent and we cannot have -runtimes -> -apps dependencies so this common module will be probably moved somewhere else inside -runtimes repo.
Btw we can change the name for sure if we find a better one :)

**Comment by danielezonca** [comment]
jenkins retest this please

**Comment by danielezonca** [comment]
jenkins retest this please

**Comment by danielezonca** [comment]
jenkins retest this please

File: addons/explainability/explainability-addon-common/src/test/java/org/kie/kogito/explainability/ExplainabilityServiceTest.java (hunk 1 above)
Comment (danielezonca) [reply addons/explainability/explainability-addon-common/src/test/java/org/kie/kogito/explainability/ExplainabilityServiceTest.java:51]: Ticket created https://issues.redhat.com/browse/KOGITO-3213

**Review by jiripetrlik — APPROVED** [review]
Looks good to me. Few minor comments.

File: addons/explainability/explainability-addon-common/src/test/java/org/kie/kogito/explainability/ExplainabilityServiceTest.java [hunk 6]
```diff
@@ -54,42 +49,37 @@
...
-                if (MODEL_NAMESPACE.equals(namespace) && MODEL_NAME.equals(name)) {
-                    return decisionModel;
-                }
-                throw new RuntimeException("Model not found.");
+
+        DecisionModels decisionModels = (namespace, name) -> {
+            if (MODEL_NAMESPACE.equals(namespace) && MODEL_NAME.equals(name)) {
+                return decisionModel;
             }
+            throw new RuntimeException("Model not found.");
```
[the remark this answers is not in the document]
Comment (danielezonca) [reply addons/explainability/explainability-addon-common/src/test/java/org/kie/kogito/explainability/ExplainabilityServiceTest.java]: Done

File: addons/explainability/explainability-quarkus-addon/src/test/java/org/kie/kogito/explainability/ApplicationMock.java [hunk 7]
```diff
@@ -14,51 +14,36 @@
...
-            public DecisionModel getDecisionModel(String namespace, String name) {
-                if (MODEL_NAMESPACE.equals(namespace) && MODEL_NAME.equals(name)) {
-                    return decisionModel;
-                }
-                throw new RuntimeException("Model not found.");
+        decisionModels = (namespace, name) -> {
+            if (Constants.MODEL_NAMESPACE.equals(namespace) && Constants.MODEL_NAME.equals(name)) {
+                return decisionModel;
             }
+            throw new RuntimeException("Model not found.");
```
[the remark this answers is not in the document]
Comment (danielezonca) [reply addons/explainability/explainability-quarkus-addon/src/test/java/org/kie/kogito/explainability/ApplicationMock.java]: Done

**Comment by danielezonca** [comment]
jenkins retest this please

**Review by r00ta — APPROVED** [review]

**Pull request merged**